cmd/hub/mcscontroller/controller.go

```diff
@@ -0,0 +1,127 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+)
+
+const (
+	serviceExportFinalizer = "lighthouse.submariner.io/service-export-finalizer"
+)
+
+// ServiceExportReconciler reconciles a ServiceExport object
+type ServiceExportReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/finalizers,verbs=get;update
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/status,verbs=get;update
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceimports,verbs=create;get;list;update;patch
+
+// Reconcile handles an update to a ServiceExport
+// By design, the event that triggered reconciliation is not passed to the reconciler
```

Comment:
but it somewhat does .. ?  you resolve the event from the NamespacedName .. I dont think we can call define this approach as fully state driven

Reply:
Added note about needing to be aware of deletion events in a description of the reconciliation logic.
the issue is that client.List won't return the deleted object so we need to explicitly delay deletion (with a finalizer) so we can retrieve the "real" name and namespace without relying on a specific name format (such as name-ns-cluster).
Changed so we do extra work on cleaning up only when DeletionTimeStamp is set, but otherwise call `reconcile` just the same for all operations.